src/Language/PureScript/TypeChecker.hs

```diff
@@ -393,11 +397,63 @@ typeCheckAll moduleName _ = traverse go
           checkOrphanInstance dictName className tys' nonOrphanModules
           let qualifiedChain = Qualified (Just moduleName) <$> ch
           checkOverlappingInstance qualifiedChain dictName className typeClass tys' nonOrphanModules
+          checkInstanceBody body
           _ <- traverseTypeInstanceBody checkInstanceMembers body
           deps'' <- (traverse . overConstraintArgs . traverse) replaceAllTypeSynonyms deps'
           let dict = TypeClassDictionaryInScope qualifiedChain idx qualifiedDictName [] className vars kinds' tys' (Just deps'')
           addTypeClassDictionaries (Just moduleName) . M.singleton className $ M.singleton (tcdValue dict) (pure dict)
           return d
+    where
+    checkInstanceBody (ViaInstanceWithDictionary viaTy _) = do
+      (tyConName, tyConArgs) <- unapplyTypeConstructor (last tys)
+        & maybe (throwError . errorMessage' ss $ ExpectedTypeConstructor className tys (last tys)) return
+      (viaTyConName, viaTyConArgs) <- unapplyTypeConstructor viaTy
+        & maybe (throwError . errorMessage' ss $ InvalidViaType className tys viaTy) return
+      kinds <- gets (types . checkEnv)
+      (kind, _) <- M.lookup tyConName kinds
+        & maybe (throwError . errorMessage' ss . CannotFindDerivingType $ disqualify tyConName) return
+      (viaKind, _) <- M.lookup viaTyConName kinds
+        & maybe (throwError . errorMessage' ss . CannotFindViaType $ disqualify viaTyConName) return
+      let tyVars = S.fromList $ usedTypeVariables (last tys)
+          viaTyVars = S.fromList $ usedTypeVariables viaTy
+          floating = NEL.nonEmpty . S.toList $ viaTyVars `S.difference` tyVars
+      FloatingViaTypeVariables className tys viaTy <$> floating
+        & maybe (return ()) (throwError . errorMessage' ss)
+      checkViaKind moduleName ss className tys viaTy
+      let saturate' = saturate $ tyVars `S.union` viaTyVars
+          a = saturate' kind (last tys) (length tyConArgs)
+          b = saturate' viaKind viaTy (length viaTyConArgs)
+          constraint = srcConstraint Coercible [] [a, b] Nothing
+          solverOptions =
+            SolverOptions
+              { solverShouldGeneralize = False
+              , solverDeferErrors      = False
+              }
+      void . flip runStateT M.empty . runWriterT $
+        entails solverOptions constraint M.empty []
```

Comment:
Does this throw away the unsolved constraints?

Reply:
Oh, but that's fine because of the current `SolverOptions`, so the solver won't emit unsolved constraints and instead just fail, right? I guess that means we can't write something like

```purescript
data T

derive via (...) instance blah :: Coercible X Y => Cls T
```

i.e. defer `Coercible` constraints in the instance head, but that's not a big loss, and maybe something that can be solved later?

Reply:
The solver indeed fails when typechecking the instance declaration if the generated constraint can’t be solved but it doesn’t try to solve the instance context, so you can write such instance.